src/main/java/com/pingcap/tikv/util/TimeUtil.java

```diff
@@ -0,0 +1,210 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.pingcap.tikv.util;
+
+/**
+ * The Time util.
+ * <p>
+ * Used for converting a timestamp to timezone-insensitive {com.pingcap.tikv.util.Date}
+ */
+public class TimeUtil {
+  /**
+   * The type Date.
+   */
+  public static class Date {
+    private int year = -1;
+    private int month = -1;
+    private int day = -1;
+
+    /**
+     * Gets year.
+     *
+     * @return the year
+     */
+    public int getYear() {
+      return year;
+    }
+
+    /**
+     * Gets month.
+     *
+     * @return the month
+     */
+    public int getMonth() {
+      return month;
+    }
+
+    /**
+     * Gets day.
+     *
+     * @return the day
+     */
+    public int getDay() {
+      return day;
+    }
+
+    /**
+     * Instantiates a new Date.
+     *
+     * @param year  the year
+     * @param month the month
+     * @param day   the day
+     */
+    public Date(int year, int month, int day) {
+      this.year = year;
+      this.month = month;
+      this.day = day;
+    }
+
+    private Date() {
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == this) {
+        return true;
+      } else if (obj instanceof Date) {
+        Date date = (Date) obj;
+        return date.year == year &&
+            date.month == month &&
+            date.day == day;
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      int v = 7;
+      v += v * 31 + year;
+      v += v * 31 + month;
+      v += v * 31 + day;
+      return v;
+    }
+
+    @Override
+    public String toString() {
+      return year + "-" + month + "-" + day;
+    }
+  }
+
+  private static int START_YEAR = 1970;
+  private static long MS_PER_DAY = 24 * 3600 * 1000;
+  private static final int[] MON_AVERAGE_YEAR = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+  private static final int[] MON_LEAP_YEAR = {31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+  private static boolean isLeapYear(int year) {
+    return year % 4 == 0 && (year % 400 == 0 || year % 100 != 0);
+  }
+
+  /**
+   * Get year from timestamp using binary search.
+   *
+   * @param timestamp the timestamp
+   * @return the result year
+   */
+  public static int getYear(long timestamp) {
+    int l = 0, r = 9999;
+    long days = timestamp / MS_PER_DAY;
+    while (l <= r) {
+      int mid = (l + r) >> 1;
+      long from1970Days = calcDays(START_YEAR, mid);
+      if (from1970Days > days) {
+        r = mid - 1;
+      } else if (from1970Days < days) {
+        l = mid + 1;
+      } else {
+        return mid + 1;
```

Comment:
why `mid + 1` should be returned when `from1970Days = days` ?

Reply:
If `from1970Days = days`, it means exactly `days` days has passed from `1970-01-01`. E.g, let `days = 365`, `mid = 1970`, thus `from1970Days = 365`, since exactly a full **365days** passed, it should be `1971-01-01 00:00:00`, so we return `mid + 1` here to represent the next year.

Reply:
well it seems that calcDays should be 
private static long calcDays(int from, int to) {          return calcDays(to) - calcDays(from);
  }
so you should return mid here.
it was my fault.